### PR TITLE
New global <fim> block within <global> in the manager's ossec.conf

### PR DESCRIPTION
**Pull request by juan70**
Hi, Team,

This PR fixes the issue https://github.com/wazuh/wazuh/issues/2378.

- Problem description
  The events coming from an agent and related to directories or files that are incidentally specified in the element `<syscheck><ignore>` on the manager are ignored, even though these files are not specified in the element `<syscheck><ignore>` on the agent.
  When not ignored by the agent, these events are sent to the manager, but they are ignored by the manager whenever the files involved are listed in `<syscheck><ignore>` on the manager.
  See `Testing before changes` below for an example.

- Solution description
  The manager's `<syscheck><ignore>` should apply only to the events/files originated on the manager itself, not those coming from an agent.
  New block, called `<fim>`, within the `<global>` section, to ignore globally the events/files coming from anywhere, i.e. manager and/or agents.
  On the manager, the `<syscheck><ignore>` elements refer now only to the files/directories of the manager itself.
  On the other hand, the `<global><fim><ignore>` elements refer to all the files/directories present in the events coming from the manager and also from the agents.
  Options `<syscheck><alert_new_files>` and `<syscheck><auto_ignore>` are moved the the new `<fim>`. They are now deprecated in the `<syscheck>` section.
  See `Testing after changes` below for an example.
   
Regards.

### CODE REVIEW

**Comment by juan70** [comment]
# Tests before changes
- Agent. Edit `ossec.log` and add:
  ```
  <syscheck>
    ...
    <directories check_all="yes" realtime="yes">/root</directories>
    ...
  </syscheck>
  ```
  (restart wazuh-agent)
  - On the agent:
  ```
  # touch /root/a.txt        => Alert on the manager: 'File '/root/a.txt' was added.'.
  ```

- Manager. Edit `ossec.log` and add:
  ```
  <syscheck>
    ...
    <ignore>/root</root>
    ...
  </syscheck>
  ```
  (restart wazuh-manager)
  - On the agent:
  ```
  # touch /root/b.txt        => No alert on the manager.
  ```
  The agent's `/root` directory is ignored, because the manager's `/root` directory is ignored. This is not correct.

**Comment by juan70** [comment]
# Tests after changes

Manager and agent installed from sources including the changes. Only the manager's behaviour is actually impacted by the changes.

Manager
-------
  - Create some directories
  ```
  # cd /root
  # mkdir -p testing/tmp1 testing/tmp/tmp{2,3,4,5}
  # tree testing
  testing/
  ├── tmp
  │   ├── tmp2
  │   ├── tmp3
  │   ├── tmp4
  │   └── tmp5
  └── tmp1
  ```
  - Edit `ossec.conf` and add:
  ```
  <global>
    ...
    <logall>yes</logall>
    ...
    <fim>
      <ignore>/root/testing/tmp/tmp2</ignore>
      <ignore>/root/testing/tmp/tmp3</ignore>
      <alert_new_files>yes</alert_new_files>
      <auto_ignore frequency="10" timeframe="3600">no</auto_ignore>
    </fim>
  </global>

  ...

  <syscheck>
    ...
    <directories check_all="yes" realtime="yes">/root/testing/tmp</directories>
    ...
    <ignore>/root/testing/tmp/tmp4</ignore>
    ...
  </syscheck>
  ```
  (`realtime="yes"`, only to have immediate feedback in `alerts.log`.)
  When the manager is restarted, these messages appear in `ossec.log`:
  ```
  # grep -E '(Ignoring|Monit).*root' /var/ossec/logs/ossec.log
  2019/02/01 10:32:15 ossec-analysisd: INFO: Ignoring file: '/root/testing/tmp/tmp2'
  2019/02/01 10:32:15 ossec-analysisd: INFO: Ignoring file: '/root/testing/tmp/tmp3'
  2019/02/01 10:32:19 ossec-syscheckd: INFO: Monitoring directory: '/root/testing/tmp', with options perm | size | owner | group | md5sum | sha1sum | sha256sum | realtime | mtime | inode.
  2019/02/01 10:32:19 ossec-syscheckd: INFO: Ignoring: '/root/testing/tmp/tmp4'
  ```



Agent
-----
  - Create the same directory structure as in the manager.
  ```
  # cd /root
  # mkdir -p testing/tmp1 testing/tmp/tmp{2,3,4,5}
  # tree testing
  testing/
  ├── tmp
  │   ├── tmp2
  │   ├── tmp3
  │   ├── tmp4
  │   └── tmp5
  └── tmp1
  ```
  - Edit `ossec.conf` and add:
  ```
  <syscheck>
    ...
    <directories check_all="yes" realtime="yes">/root/testing/tmp</directories>
    ...
    <ignore>/root/testing/tmp/tmp5</ignore>
    ...
  </syscheck>
  ```
  When the agent is restarted, these messages appears in `ossec.log`:
  ```
  # grep -E '(Ignoring|Monit).*root' /var/ossec/logs/ossec.log
  2019/02/01 10:32:58 ossec-syscheckd: INFO: Monitoring directory: '/root/testing/tmp', with options perm | size | owner | group | md5sum | sha1sum | sha256sum | realtime | mtime | inode.
  2019/02/01 10:32:58 ossec-syscheckd: INFO: Ignoring: '/root/testing/tmp/tmp5'
  ```



Testing
-------
  - On the agent:
  ```
  # touch /root/testing/tmp1/a.txt        => no alert. Directory not monitored by syscheck on the agent.
  # touch /root/testing/tmp/tmp2/a.txt    => no alert. Directory ignored by the manager (`fim_ignore`).
  # touch /root/testing/tmp/tmp3/a.txt    => no alert. Likewise.
  # touch /root/testing/tmp/tmp4/a.txt    => alert. Parent directory monitored (`tmp`) on the agent. Not ignored on the manager (not in `fim_ignore`).
  # touch /root/testing/tmp/tmp5/a.txt    => no alert. Directory ignored on the agent.
  ```
  Same results for `rm` of these same files, for the same reasons.

  - On the manager:
  ```
  # touch /root/testing/tmp1/a.txt        => no alert. Directory not monitored by syscheck on the manager.
  # touch /root/testing/tmp/tmp2/a.txt    => no alert. Directory ignored by the manager (`fim_ignore`).
  # touch /root/testing/tmp/tmp3/a.txt    => no alert. Likewise.
  # touch /root/testing/tmp/tmp4/a.txt    => no alert. Directory ignored by syscheck on the manager.
  # touch /root/testing/tmp/tmp5/a.txt    => alert. Directory not ignored in any way.
  ```
  Same results for `rm` of these same files, for the same reasons.

Some more tests have been carried out:
- An installation from sources creates the correct `ossec.conf` for the manager and for the agent.
  - On the manager: `<fim> ... </fim>` block containing `<alert_new_files>` and `<auto_ignore>`.
  - On the manager and the agent, `<alert_new_files>` and `<auto_ignore>` are not present within the `<syscheck>` section (they are deprecated). If they are present, a warning message is sent to `ossec.log`.
- The manager starts correctly when the `<fim></fim>` block is completely empty or is not at all present.
- An error occurs whenever the `<fim>...</fim>` block is outside of the `<global>` section. 
- `<alert_new_files>` and `<auto_ignore>` must contain a correct value (`yes` or `no`, not empty).
- The `<ignore>` element is not required. And, if some are present, they can be empty.
- With `<auto_ignore frequency="5" timeframe="60">yes</auto_ignore>`, the alerts are indeed limited to 5 within 1 minute. The 6th event for a change in a file is not alerted. After one minute, new events are alerted again.
- `<alert_new_files>` and `<auto_ignore>` in the manager's `<syscheck>` section are deprecated, but they are still functional for the time being.



API calls
---------
Sample from the `ossec.conf` (only the `<global>` section):
```
  ...
  <global>
    <jsonout_output>yes</jsonout_output>
    <alerts_log>yes</alerts_log>
    <logall>yes</logall>
    <logall_json>yes</logall_json>
    <email_notification>no</email_notification>
    <smtp_server>smtp.example.wazuh.com</smtp_server>
    <email_from>ossecm@example.wazuh.com</email_from>
    <email_to>recipient@example.wazuh.com</email_to>
    <email_maxperhour>12</email_maxperhour>
    <email_log_source>alerts.log</email_log_source>
    <queue_size>131072</queue_size>
    <fim>
      <ignore>/root/testing/tmp/tmp2</ignore>
      <ignore>/root/testing/tmp/tmp3</ignore>
      <alert_new_files>yes</alert_new_files>
      <auto_ignore frequency="10" timeframe="3600">no</auto_ignore>
    </fim>
  </global>
 ....
```
```
# curl -u foo:bar -X GET "http://localhost:55000/agents/000/config/analysis/global?pretty"
{
   "error": 0,
   "data": {
      "global": {
         "email_notification": "no",
         "max_output_size": 0,
         "alerts_log": "yes",
         "zeromq_output": "no",
         "host_information": 8,
         "jsonout_output": "yes",
         "rotate_interval": 0,
         "rootkit_detection": 8,
         "integrity_checking": 8,
         "memory_size": 8192,
         "logall": "yes",
         "prelude_output": "no",
         "stats": 4,
         "fim": {
            "ignore": [
               "/root/testing/tmp/tmp2",
               "/root/testing/tmp/tmp3"
            ],
            "auto_ignore": {
               "item": "no",
               "frequency": 10,
               "timeframe": 3600
            },
            "alert_new_files": "yes"
         },
   ...
```

When the `<fim>` block is not present at all in `ossec.conf`:
```
   ...
         "logall": "yes",
         "prelude_output": "no",
         "stats": 4,
         "fim": {
            "auto_ignore": {
               "item": "no",
               "frequency": 10,
               "timeframe": 3600
            },
            "alert_new_files": "yes"
         },
   ...
```
The options `auto_ignore` and `alert_new_files` are shown because they have default values.


Nevertheless, `curl -u foo:bar -X GET "http://localhost:55000/manager/configuration?section=global&pretty"` fails  to show the `fim` block correctly, because some specific development must still be done in the framework.

File: src/config/global-config.c [hunk 1]
```diff
@@ -117,7 +122,113 @@ int Read_GlobalSK(XML_NODE node, void *configp, __attribute__((unused)) void *ma
...
+        if (!node[i]->element) {
+            merror(XML_ELEMNULL);
+            return (OS_INVALID);
+        } else if (!node[i]->content) {
+            merror(XML_VALUENULL, node[i]->element);
+            return (OS_INVALID);
+
+        } else if (strcmp(node[i]->element, xml_ignore) == 0) {
+             if (Config) {
+                 fim_ign_size++;
```
Comment (chemamartinez) [inline src/config/global-config.c]: Sometimes we get undesired content in `Config->fim_ignore`:

```
2019/02/05 06:12:33 ossec-analysisd: INFO: Ignoring file: '  �arU'
2019/02/05 06:12:33 ossec-analysisd: INFO: Ignoring file: '�arU'
```

File: src/config/global-config.c [hunk 2]
```diff
@@ -55,6 +56,8 @@ int Read_GlobalSK(XML_NODE node, void *configp, __attribute__((unused)) void *ma
             merror(XML_VALUENULL, node[i]->element);
             return (OS_INVALID);
         } else if (strcmp(node[i]->element, xml_auto_ignore) == 0) {
+            mwarn("The <syscheck><%s> tag is deprecated, please use <global><fim><%s> instead.", xml_auto_ignore, xml_auto_ignore);
```
Comment (chemamartinez) [inline src/config/global-config.c]: Warnings about deprecated options are being logged for three times each start-up:

```
2019/02/05 07:01:46 ossec-testrule: WARNING: The <syscheck><auto_ignore> tag is deprecated, please use <global><fim><auto_ignore> instead.
2019/02/05 07:01:46 ossec-maild: WARNING: The <syscheck><auto_ignore> tag is deprecated, please use <global><fim><auto_ignore> instead.
2019/02/05 07:01:46 ossec-analysisd: WARNING: The <syscheck><auto_ignore> tag is deprecated, please use <global><fim><auto_ignore> instead.

**Pull request closed without merging**